cmd/kops/toolbox_instance_selector.go

```diff
@@ -0,0 +1,487 @@
+	}
+	if flags[instanceGroupCount] != nil {
+		opts.InstanceGroupCount = *commandline.IntMe(flags[instanceGroupCount])
+	}
+	return opts
+}
+
+func getClusterZones(subnets []kops.ClusterSubnetSpec) ([]string, error) {
+	region := ""
+	zones := []string{}
+	for _, subnet := range subnets {
+		zoneRegion := subnet.Zone[:len(subnet.Zone)-1]
+		zones = append(zones, subnet.Zone)
+		if region != "" && zoneRegion != region {
+			return nil, fmt.Errorf("clusters cannot span multiple regions")
+		}
+		region = zoneRegion
+	}
+	if len(zones) == 0 {
+		return nil, fmt.Errorf("the cluster must include at least 1 subnet")
+	}
+	return zones, nil
+}
+
+func createInstanceGroup(groupName, clusterName string, zones []string) *kops.InstanceGroup {
+	ig := &kops.InstanceGroup{}
+	ig.ObjectMeta.Name = groupName
+	ig.Spec.Role = kops.InstanceGroupRoleNode
+	ig.Spec.Subnets = zones
+	ig.ObjectMeta.Labels = make(map[string]string)
+	ig.ObjectMeta.Labels[kops.LabelClusterName] = clusterName
+
+	ig.AddInstanceGroupNodeLabel()
+	return ig
+}
+
+func decorateWithInstanceGroupSpecs(instanceGroup *kops.InstanceGroup, instanceGroupOpts InstanceSelectorOptions) *kops.InstanceGroup {
+	ig := instanceGroup
+	ig.Spec.MinSize = &instanceGroupOpts.NodeCountMin
+	ig.Spec.MaxSize = &instanceGroupOpts.NodeCountMax
+	ig.Spec.RootVolumeSize = instanceGroupOpts.NodeVolumeSize
+	ig.Spec.AdditionalSecurityGroups = instanceGroupOpts.NodeSecurityGroups
+	return ig
+}
+
+func decorateWithMixedInstancesPolicy(instanceGroup *kops.InstanceGroup, usageClass string, instanceSelections []string) (*kops.InstanceGroup, error) {
+	ig := instanceGroup
+	ig.Spec.MachineType = instanceSelections[0]
+
+	if usageClass == usageClassSpot {
+		ondemandBase := int64(0)
+		ondemandAboveBase := int64(0)
+		spotAllocationStrategy := "capacity-optimized"
+		ig.Spec.MixedInstancesPolicy = &kops.MixedInstancesPolicySpec{
+			Instances:              instanceSelections,
+			OnDemandBase:           &ondemandBase,
+			OnDemandAboveBase:      &ondemandAboveBase,
+			SpotAllocationStrategy: &spotAllocationStrategy,
+		}
+	} else if usageClass == usageClassOndemand {
+		ig.Spec.MixedInstancesPolicy = &kops.MixedInstancesPolicySpec{
+			Instances: instanceSelections,
+		}
+	} else {
+		return nil, fmt.Errorf("error node usage class not supported")
+	}
+
+	generatedWithLabelKey := "kops.k8s.io/instance-selector"
+	if ig.Spec.CloudLabels == nil {
+		ig.Spec.CloudLabels = make(map[string]string)
+	}
+	ig.Spec.CloudLabels[generatedWithLabelKey] = "1"
+
+	return ig, nil
+}
+
+func decorateWithClusterAutoscalerLabels(instanceGroup *kops.InstanceGroup) *kops.InstanceGroup {
+	ig := instanceGroup
+	clusterName := instanceGroup.ObjectMeta.Name
+	if ig.Spec.CloudLabels == nil {
+		ig.Spec.CloudLabels = make(map[string]string)
+	}
+	ig.Spec.CloudLabels["k8s.io/cluster-autoscaler/enabled"] = igName
+	ig.Spec.CloudLabels["k8s.io/cluster-autoscaler/"+clusterName] = igName
```

Comment:
is `igName` the best value for these ASG tags? I think I've used `1` in the past but I dont know if there's a standardized value.

Reply:
It does seem like 1 is more prominently used. The CA docs says the key is the only thing it looks at, but I'll change to use "1" as the tag value to use the more familiar value.